nvdaHelper/localWin10/oneCoreSpeech.cpp

```diff
@@ -22,99 +21,393 @@ This license can be found at:
+void preventEndUtteranceSilence_(std::shared_ptr<winrtSynth> synth);
+
+void* __stdcall ocSpeech_initialize(ocSpeech_Callback fn) {
+	LOG_INFO(L"ocSpeech_initialize");
+	// Ensure there are no pending shared locks, all async speak calls must be finished
+	auto lock = getUniqueLock_(L"ocSpeech_initialize");
+	if (!lock) {
+		return nullptr;
+	}
+	LOG_DEBUG(L"ocSpeech_initialize lock acquired");
+
+	auto token = g_state.activate(fn);
+	auto synth = g_state.getSynth(token);
+	if (!synth) {
+		LOG_ERROR(L"Unable to initialize.");
+		return nullptr;
 	}
+	preventEndUtteranceSilence_(synth);
+	return token;
 }
 
-OcSpeech* __stdcall ocSpeech_initialize() {
-	auto instance = new OcSpeech;
-	return instance;
+void __stdcall ocSpeech_terminate(void* token) {
+	std::wstringstream ss;
+	ss << L"ocSpeech_terminate, token: " << token;
+	LOG_INFO(ss.str());
+	// Acquire a unique lock to ensure that no callbacks can happen while changing state.
+	// Changing state must be atomic.
+	auto lock = getUniqueLock_(ss.str());
+	if (!lock) {
+		return;
+	}
+	if (!g_state.isTokenValid(token)) {
+		LOG_ERROR(L"ocSpeech_terminate error");
+		return;
+	}
+	g_state.terminate();
 }
 
-void __stdcall ocSpeech_terminate(OcSpeech* instance) {
-	delete instance;
+
+/* Internal convenience pairing for a winRT synth result.
+*/
+struct SpeakResult {
+	Buffer buffer;
+	std::wstring markersStr;
+};
+
+/*Call the external callback.
+* This requires a shared lock on the callback / g_OcSpeechStateMutex.
+* This shared ownership prevents (blocks) unique locks from being acquired, thus blocking
+* ocSpeech_initialize and ocSpeech_terminate.
+* If the given token is no longer valid, the callback will not be called.
+*/
+void protectedCallback_(
+	void* token,
+	std::optional<SpeakResult> result,
+	std::function<ocSpeech_CallbackT> cb
+) {
+	// Prevent any unique locks being acquired.
+	SharedLock lock(g_OcSpeechStateMutex, std::defer_lock);
+	const bool owned = lock.try_lock();
+	if (!owned) {
+		LOG_DEBUG(L"protectedCallback_ - Unable to lock. Token: " << token);
+		return;
+	}
+	LOG_DEBUG(L"protectedCallback, token: " << token);
+	if (!g_state.isTokenValid(token)) {
+		LOG_INFO(L"not calling CB, token: " << token);
+		return;
+	}
+	LOG_DEBUG(L"calling CB, token: " << token);
+	if (result.has_value()) {
+		cb(result->buffer.data(), result->buffer.Length(), result->markersStr.c_str());
+	}
+	else {
+		cb(nullptr, 0, nullptr);
+	}
+	LOG_DEBUG(L"Finished CB, token: " << token);
 }
 
-void __stdcall ocSpeech_setCallback(OcSpeech* instance, ocSpeech_Callback fn) {
-	instance->setCallback(fn);
+bool isUniversalApiContractVersion_(const int major, const int minor);
+
+bool __stdcall ocSpeech_supportsProsodyOptions() {
+	return isUniversalApiContractVersion_(5, 0);
 }
 
-void OcSpeech::setCallback(ocSpeech_Callback fn) {
-	callback = fn;
+std::wstring createMarkersString_(IVectorView<IMediaMarker> markers) {
+	std::wstring markersStr;  // for large strings, reserving would speed this up.

```

Comment:
should this be reversed then? I assume short speech is fast enough regardless and we want to optimize against the worst case here.

Reply:
I suggest we do this in another PR to master

Reply:
I probably should have said "may speed this up". To be honest, I don't think optimising this is likely to be worth the time. There may be nanoseconds to save. I'd certainly recommend benchmarking and profiling the code first. This is running async to NVDA, so not blocking other tasks. As long as the new synth data is returning before the last audio is finished playing then that aspect is not user observable. However, a delay here may be observed by a user as latency on the synths "cold start".